debugger/src/debugger.rs

```diff
@@ -0,0 +1,1111 @@
+use crate::dap_types::*;
+use crate::debug_adapter::DapStatus;
+use crate::debug_adapter::*;
+
+use crate::DebuggerError;
+use anyhow::{anyhow, Result};
+use capstone::{arch::arm::ArchMode, prelude::*, Capstone, Endian};
+use probe_rs::config::TargetSelector;
+use probe_rs::debug::DebugInfo;
+use probe_rs::flashing::{download_file, download_file_with_options, DownloadOptions, Format};
+use probe_rs::{Core, CoreStatus, MemoryInterface, Probe, Session, WireProtocol};
+use serde::Deserialize;
+use std::{
+    env::{current_dir, set_current_dir},
+    io,
+    io::{Read, Write},
+    net::{Ipv4Addr, TcpListener, ToSocketAddrs},
+    path::PathBuf,
+    str::FromStr,
+    thread,
+    time::{Duration, Instant},
+};
+use structopt::StructOpt;
+
+fn parse_protocol(src: &str) -> Result<WireProtocol, String> {
+    WireProtocol::from_str(src)
+}
+fn default_wire_protocol() -> Option<WireProtocol> {
+    Some(WireProtocol::Swd)
+}
+fn default_console_log() -> Option<ConsoleLog> {
+    Some(ConsoleLog::Error)
+}
+fn parse_console_log(src: &str) -> Result<ConsoleLog, String> {
+    ConsoleLog::from_str(src)
+}
+/// The level of information to be logged to the debugger console
+#[derive(Copy, Clone, PartialEq, Debug, serde::Serialize, serde::Deserialize)]
+pub enum ConsoleLog {
+    Error,
+    Info,
+    Debug,
+} //TODO: It would be nice instead to tap into log write once the DebugAdapter has been initialized, and intercept RUST like log info
+impl std::str::FromStr for ConsoleLog {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match &s.to_ascii_lowercase()[..] {
+            "error" => Ok(ConsoleLog::Error),
+            "info" => Ok(ConsoleLog::Info),
+            "debug" => Ok(ConsoleLog::Debug),
+            _ => Err(format!(
+                "'{}' is not a valid console log level. Choose from [error, info, debug].",
+                s
+            )),
+        }
+    }
+}
+
+//TODO: When we implement multi-threaded server, this static handling needs work
+pub static mut LAST_KNOWN_STATUS: CoreStatus = CoreStatus::Unknown;
+
+/// Shared options for all commands which use a specific probe
+#[derive(StructOpt, Clone, Deserialize, Debug, Default)]
+pub struct DebuggerOptions {
+    /// Path to the requested working directory for the debugger
+    #[structopt(long, parse(from_os_str), conflicts_with("dap"))]
+    cwd: Option<PathBuf>,
+
+    /// Binary to debug as a path. Relative to `cwd`, or fully qualified.
+    #[structopt(long, parse(from_os_str), conflicts_with("dap"))]
+    program_binary: Option<PathBuf>,
+
+    /// The number associated with the debug probe to use. Use 'list' command to see available probes
+    #[structopt(long = "probe-index", default_value)]
+    #[serde(default)]
+    pub(crate) probe_index: usize,
```

Comment:
Please use an `Option` here, otherwise this leads to problem when multiple probes are connected. Right now this defaults to 0, and seems to just open the first probe found, even when multiple probes are connected.

Reply:
Also, I would like to add, that we have moved to a better way of specifying the debug probe which is deterministic and not dependant on USB scan order for our other tools. E.g. https://github.com/probe-rs/cargo-flash/blob/master/src/main.rs#L74

Reply:
Ouch. Changing this impacts the VSCode extension also, but I can see that it is worth it.